File: CoinChange.java
class Main {
  public static void main(String[] args) {
    int[] coins = {1, 2, 5};
    int amount = 11;
    System.out.println(coinChange(coins, amount));
  }

  public static int coinChange(int[] coins, int amount) {
        int numCoins = coins.length;
        int min = Integer.MAX_VALUE - 1;
        int curr = min;
        int[] dp = new int[amount + 1];
        dp[0] = 0;

        for(int i = 1; i <= amount; i++){
          min = Integer.MAX_VALUE;
            for(int j = 0; j < numCoins; j++){
                int coinVal = coins[j];
                if(coinVal <= i){
                  curr = 1 + dp[i - coinVal];
                }
                if(curr < min){
                    min = curr;
                }
            }
          dp[i] = min;
        }

        return dp[amount];
    }
}

File: CoinChange.cpp
#include <iostream>
#include <fstream>
using namespace std;

int coinChange(int total, int coins[]);

int main() {
  int total;
  int numCoins;
  int coins[numCoins];

  fstream file("coins.txt");
  file >> total;
  file >> numCoins;

  for(int i = 0; i < numCoins; i++){
    file >> coins[i];
  }

  cout << coinChange(total, coins) << endl;
}

int coinChange(int total, int coins[]){
  int dp[total + 1];
  int curr;
  int min;

  dp[0] = 0;

  for(int i = 1; i <= total; i++){
    min = INT32_MAX;
    for(int j = 0; j < sizeof(coins) / sizeof(int) + 1; j++){
      if(coins[j] <= i){
        curr = 1 + dp[i - coins[j]];
      }
      if(curr < min){
        min = curr;
      }
    }

    dp[i] = min;
  }
  
  return dp[total];
}
